First try at making a generic guide book.

Bernasss12 opened the pull request:
- Added simple Item, still have to attach it to the event listener and add a recipe but that's not a priority.
- Added a scale-able GUI that works around the already used translation based book text.
- Semi-implemented slider functionality but that's also not a priority as of now.

The purpose of this pull request is to get as much feedback as possible, but note I'm aware that a lot of the calculations still need to be cleaned up, as of now everything is just working but is by no means perfect.

Also I just based on mc's source code so there must be better ways to do most of the things I did, so go at it I want to improve this (and myself).

Any questions please feel free to ask away!

Commit:
First try at making a generic guide book.

Added simple Item, still have to attach it to the event listener and add a recipe but that's not a priority.
Added a scale-able GUI that works around the already used translation based book text.
Semi-implemented slider functionality but that's also not a priority as of now.

Bernasss12 commented:
Some screenshots since I forgot before:

![image](https://user-images.githubusercontent.com/6233500/59111530-1af64880-8939-11e9-8774-326259816cf8.png)

![image](https://user-images.githubusercontent.com/6233500/59111558-23e71a00-8939-11e9-9553-5f017fb8f862.png)

Edit, another one:

![image](https://user-images.githubusercontent.com/6233500/59111626-4bd67d80-8939-11e9-86d2-361b81e5ac45.png)

Bernasss12 commented:
As can be seen in the latter screenshot the line transitions can be improved... I have a couple ideas but don´t have time to implement right now

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,171 @@
+		mc.renderEngine.bindTexture(GUI_BOOK_TEXTURE);
+		int xRem = x;
+		int yRem = y;
+		int eight = 8;
```
Instead of "magic constants" name this for what it actually does; maybe something like `scaleFactor` from what I can see.

Bernasss12 replied:
Done

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,171 @@
+	public void drawGuiForeGroundLayer(){
+		GlStateManager.pushMatrix();
+		GlStateManager.translate(1.0F, 1.0F, 10.0F);
+		drawCenteredString(this.fontRenderer, StringUtils.translate("book.quickstart.title").trim(), this.width / 2, (((-((this.height/4)*3)/8)*4)+this.height/2) + 4, 16777215);
```
instead of `16777215` use `0xFFFFFF` (the same number but in hex format), it's more readable in this case since it's a color value

Bernasss12 replied:
Yes makes sense, this was just what was used in the default mc code

Bernasss12 replied:
Fixed

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,171 @@
+		GlStateManager.pushMatrix();
+		GlStateManager.translate(1.0F, 1.0F, 10.0F);
+		drawCenteredString(this.fontRenderer, StringUtils.translate("book.quickstart.title").trim(), this.width / 2, (((-((this.height/4)*3)/8)*4)+this.height/2) + 4, 16777215);
+		drawCenteredString(this.fontRenderer, "Page " + PAGE + " out of " + PAGE_MAX, this.width / 2, this.height - 25, 16777215);
```
same as above; `16777215` → `0xFFFFFF`

Bernasss12 replied:
Fixed

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,171 @@
+					line = sentence;
+					sentence = "";
+				}
+				drawString(this.fontRenderer, line, (((-((this.width/4)*3)/8)*4)+this.width/2) + 24, (((-((this.height/4)*3)/8)*4)+this.height/2) + 24 + (10 * line_number++), 16777215);
```
same as above; `16777215` → `0xFFFFFF`

the width and height calculations look refactorable (`(((-((x/4)*3)/8)*4)+x/2) + 24` = `(int) (-23.5 * x) + 24`), but I don't know if they're equivalent or if this makes use of int division. Try it, though. Or at least remove some of those parens, Java has order of operations

Bernasss12 replied:
I know I have to spend some time simplifying these calculations I was just adding things as needed

Bernasss12 replied:
This equates to:
`x * 1/8 + 24`

dblsaiko replied:
Oh whoops, yes that's correct. Make sure to use `x * 0.125f` or `x * 1/8f` though or it will do integer division

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,171 @@
+	@Override
+	public void drawScreen(int mouseX, int mouseY, float partialTicks) {
+
+		this.drawGuiBackgroundLayer(((this.width/4)*3)/8, ((this.height/4)*3)/8);
```
same here, refactor expressions

Bernasss12 replied:
Same here for me

Bernasss12 replied:
And this equates to:
`x * 3/32`

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,171 @@
+	//Book vars
+	private int PAGE = 1;
+	private int PAGE_MAX = 10;
+	private int MAX_LENGTH = 50;
```
only constants (static+final) should be title case, make these var names camel case

Bernasss12 replied:
Alright done

dblsaiko commented on common/logisticspipes/items/ItemGuideBook.java:
```diff
@@ -0,0 +1,25 @@
+	@Override
+	public ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer player, EnumHand hand) {
+		ItemStack itemstack = player.getHeldItem(hand);
+		if(!world.isRemote){return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);}
```
`new ActionResult<ItemStack>` → `new ActionResult<>`

Bernasss12 replied:
Hahaha IDEA has the same opinion
![image](https://user-images.githubusercontent.com/6233500/59119121-e5a62680-8949-11e9-8caf-2306ebdb7b72.png)

Fixed

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,171 @@
+		GlStateManager.translate(this.width / 2, this.height / 2, 0.0F);
+		/**TopLeft*/
+		drawTexturedModalRect(0, 0, 0, 0, eight * 4, eight * 4);
+		xRem -= 4;
```
Instead of doing this here, just set `xRem = x - 8` and `yRem = y - 8` initially?

Bernasss12 replied:
Yes it is the same and I just did it this way as a reminder as to why I have to do it at all

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,171 @@
+			for(int i = 0; i < xRem; i++){
+				drawTexturedModalRect(eight * (i + 4), eight * (j + 4), eight * 3, eight * 3, eight * 2, eight * 2);
+			}
+		}
```
All this drawing can be replaced by 9 draw calls only: 4 for the edges, 4 for the sides (stretch the texture horizontally or vertically depending on the side) and 1 for the center (stretch in both directions)

However, that will cause the texture to be stretched, obviously. But that shouldn't matter in this case since it doesn't have any variation (such as noise) in it.

Bernasss12 replied:
Just tell me how to strech it because I didn't use it just because I didn´t know how to

dblsaiko replied:
`drawScaledCustomSizeModalRect(x, y, u, v, uWidth, vHeight, width, height, textureWidth, textureHeight)`

where x,y,width,height are position on screen, and u,v,uWidth,uHeight are texture coordinate related (these are in pixels, when you set textureWidth and textureHeight correctly)

(and obviously replace textureWidth and textureHeight with the dimensions of the texture file)

Should work.

Bernasss12 replied:
Unless you meant for me to use `GlStateManager.scale()`

dblsaiko reviewed (commented):
I know formatting inconsistencies are all over the whole codebase, but format your code before committing (don't do that in already existing classes though because they probably have other formatting all over the place already). Shortcut to do that in IDEA is apparently ctrl+alt+l (I use different bindings than the default so I don't know if that's actually the correct one)

(I still want to do a "format everything" commit at some point :P)

Also, unless you'll implement "correct splitting" (I guess that would be splitting at syllables? No idea how you actually split words correctly) I'd rather split at word bounds (i.e. spaces, after commas, anything that isn't `[A-Za-z0-9]`, probably), that way it doesn't look weird

Bernasss12 commented:
> Also, unless you'll implement "correct splitting" (I guess that would be splitting at syllables? No idea how you actually split words correctly) I'd rather split at word bounds (i.e. spaces, after commas, anything that isn't `[A-Za-z0-9]`, probably), that way it doesn't look weird

I could try to implement that.
I wasn't thinking about splitting at syllables although that would be the correct way.

What I mean was that sometimes only one character carries to the next line when it could just no have been replaced by the dash in the first place, the other occasion I thought could be better was when commas and punctuation in general carries to a new line, again being replaced by an unnecessary dash in the previous one.

Commit:
Resolved multiple issues

- Auto-indented before committing.
- Reworked background drawing
- Fixed everything I was asked for :P

Commit:
Added slider functionality changed some things. 

- Fully implemented the slider functionality but still has no uses as no sentence fills the whole space.
- Changed the drawing functions to a more flexible one, visually the same.
- Everything should scale better now.
- Still waiting for more feedback :P

Bernasss12 commented:
Well, I believe this is going on well, please comment on your thoughts.

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -188,4 +229,12 @@ protected void mouseReleased(int mouseX, int mouseY, int state) {
+	public void onGuiClosed() {
+		NBTTagCompound nbtTagCompound = book.getTagCompound();
+		nbtTagCompound.setInteger("page", page);
+		nbtTagCompound.setFloat("sliderProgress", sliderProgress);
```
This won't work in anything but creative mode since the client can't modify item data. You'll probably have to write some kind of netcode (packets) to tell the server to change the item data

Bernasss12 replied:
Ok I'll have a look at that

Bernasss12 replied:
Fixed in the latest commit

Commit:
Many changes

- Packet implementation (sort of)
- Make working slider, check it out.
- Created entry nr 11 on .lang file, used only to debug.

Bernasss12 commented:
As of now the NBT stored page is not working as well as the semi-transparent part of the gui is not blending right

Commit:
I think this is a good enough version for evaluation

- All basic functionality of a normal book (and more) are implemented, now onto the good part.

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,312 @@
+
+import org.lwjgl.opengl.GL11;
+import org.lwjgl.opengl.GLSync;
+import scala.actors.threadpool.Arrays;
```
java.util.Arrays

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,312 @@
+		/**Getting information from item NBT*/
+		book = mc.player.inventory.getStackInSlot(itemSlot);
+		if (book.hasTagCompound()) {
+			System.out.println("Book has nbt");
```
Are these debug prints? Remove before merging

Bernasss12 replied:
Removed

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,312 @@
+		NBTTagCompound nbtTagCompound = new NBTTagCompound();
+		nbtTagCompound.setInteger("page", page);
+		nbtTagCompound.setFloat("sliderProgress", sliderProgress);
+		ItemGuideBook.setCurrentPage(itemSlot, mc.player, nbtTagCompound);
```
That's not what I was saying, you're still kinda allowing setting arbitrary NBT. Make setCurrentPage take `page` and `sliderProgress` instead of a NBT tag, and send those instead of the NBT over the network.

Bernasss12 replied:
Ok it shouldn't be a problem, why is it better though?

dblsaiko commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -0,0 +1,54 @@
+
+	@Getter
+	@Setter
+	private NBTTagCompound nbt;
```
With this you're allowing the client to set arbitrary NBT on the item, which is bad. Send `page` and `sliderProgress` instead of a NBT tag.

dblsaiko commented on common/logisticspipes/items/ItemGuideBook.java:
```diff
@@ -0,0 +1,38 @@
+		nbt.setFloat("sliderProgress", 0.0F);
+	}
+
+	public static void setCurrentPage(int slot, EntityPlayer player, NBTTagCompound nbt) {
```
Take `page` and `sliderProgress` instead of an arbitrary NBT tag

dblsaiko commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -0,0 +1,54 @@
+	@Override
+	public void processPacket(EntityPlayer player) {
+		ItemStack book = player.inventory.getStackInSlot(slot);
+		book.setTagCompound(nbt);
```
Instead of overriding the entire tag (which would clear stuff like custom naming) create the NBT if it doesn't exist, and then set `page` and `sliderProgress` on the existing tag

Also check if the item is actually a book, you don't want to allow setting NBT on unrelated items

Bernasss12 replied:
So the packet has to get the `slot`, the `EnumHand`, the `page `and the `sliderProgress`?

Bernasss12 replied:
If not the `EnumHand` some kind of boolean to be able to check the item, am I wrong?

dblsaiko commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -0,0 +1,54 @@
+
+	@Getter
+	@Setter
+	private int slot;
```
This will break when the book is in the off hand, since that's it's own inventory and with this code you can't distinguish between slot 0 of the offhand and slot 0 of the main inventory.

Bernasss12 replied:
Fixed and tested

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,312 @@
+	private int maxLines;
+
+	//Text vars
+	List<String> text = new ArrayList<>();
```
`private`

Bernasss12 replied:
Riiiight forgot

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,312 @@
+			this.page = nbtTagCompound.getInteger("page");
+			this.sliderProgress = nbtTagCompound.getFloat("sliderProgress");
+		} else {
+			System.out.println("Book doesn't have NBT");
```
Again, remove before merging

Bernasss12 replied:
Removed

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,312 @@
+				slider.y = mouseY - slider.height / 2;
+				sliderProgress = (slider.y - sliderTopY) / (float) (sliderBotY - sliderTopY);
+			}
+			System.out.println(sliderProgress);
```
Again, remove before merging

Bernasss12 replied:
Removed

dblsaiko commented on common/logisticspipes/items/ItemGuideBook.java:
```diff
@@ -0,0 +1,38 @@
+		ItemStack itemstack = player.getHeldItem(hand);
+		if (!world.isRemote) {return new ActionResult<ItemStack>(EnumActionResult.SUCCESS, itemstack);}
+		player.openGui(LogisticsPipes.instance, GuiIDs.GUI_Guide_Book_ID, world, player.getPosition().getX(), player.getPosition().getY(), player.getPosition().getZ());
+		return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
```
Duplicate code, replace with

```java
ItemStack itemstack = player.getHeldItem(hand);
if (world.isRemote)
  player.openGui(LogisticsPipes.instance, GuiIDs.GUI_Guide_Book_ID, world, player.getPosition().getX(), player.getPosition().getY(), player.getPosition().getZ());
return new ActionResult<>(EnumActionResult.SUCCESS, itemstack);
```

Bernasss12 replied:
Replaced

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,312 @@
+	private static final ResourceLocation GUI_BOOK_TEXTURE = new ResourceLocation(LPConstants.LP_MOD_ID, "textures/gui/guide_book.png");
+
+	private GuiButton nextPageBtn, prevPageBtn, slider;
+	private int itemSlot;
```
This never gets set anywhere, does it? Add it as a constructor parameter

Bernasss12 replied:
I ended up not really needing it at all
Edit: I do need it, but I just found out that the variable is not being set anywhere and it is used twice and I haven't a clue how that's working

dblsaiko commented on common/logisticspipes/network/GuiHandler.java:
```diff
@@ -261,6 +262,10 @@ public Object getClientGuiElement(int ID, EntityPlayer player, final World world
 		}
 
+		if (ID == GuiIDs.GUI_Guide_Book_ID) {
+			return new GuiGuideBook();
```
Hmm, don't do the GUI construction here, you won't be able to pass the additional parameters that you will need. You could use the x,y,z coordinates for extra data if you really want to, though, instead of the player's coordinates, since they're not actually used for anything.

Bernasss12 replied:
Where should I make it? I thought I had to do it here

dblsaiko replied:
I'd add a method to the client/common proxy (don't construct it directly from the item class) that opens the gui (obviously only on the client side, on the server it's a no-op), and call that from the onItemUse method instead of this one

Bernasss12 replied:
I couldn't make this elsewhere, but as I read that the x y z are not needed to be actual coordinates but could be any arbitrary value I am using the x value as a way to pass the hand through to the Gui.
Until I find a proper solution I think this will have to be :(

Commit:
Remade some key parts

- Routed the GUI opening through the ClientProxy.
- Reformatted the packet to not use NBT but instead simple values.
- Made it work _flawlessly_ with the offhand.

dblsaiko commented on common/logisticspipes/proxy/side/ClientProxy.java:
```diff
@@ -244,12 +248,21 @@ public void openFluidSelectGui(final int slotId) {
+		EntityPlayer player = mc.player;
+		World world = mc.world;
+		BlockPos pos = player.getPosition();
+		player.openGui(LogisticsPipes.instance, guiId, world, hand.ordinal(), pos.getY(), pos.getZ());
```
What about `mc.displayGuiScreen(new GuiGuideBook(hand))` instead of `openGui`? You're overcomplicating this way too much here. Then you can also remove all the GuiHandler stuff...

Bernasss12 replied:
Corrected

dblsaiko commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -0,0 +1,76 @@
+	public void processPacket(EntityPlayer player) {
+		ItemStack book;
+		if (hand == EnumHand.MAIN_HAND) {
+			if (!(player.getHeldItemMainhand().getItem().getClass() == ItemGuideBook.class)) return;
```
No. Check against the item instance that you registered, not the class, or anything else. ~~You actually have to save it somewhere when registering, though, which you currently aren't doing.~~ Nevermind, there's an object holder for it.

Bernasss12 replied:
`if(book.getItem() != LPItems.itemGuideBook) return;` Would this be the right way?

bziemons replied:
@Bernasss12 yes

Bernasss12 replied:
Great!

dblsaiko commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -0,0 +1,76 @@
+			if (!(player.getHeldItemMainhand().getItem().getClass() == ItemGuideBook.class)) return;
+			book = player.getHeldItemMainhand();
+		} else {
+			if (!(player.getHeldItemOffhand().getItem().getClass() == ItemGuideBook.class)) return;
```
See above

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,316 @@
+
+	//Drawing Constants
+	private final float textureSide$x = 256.0F;
+	private final float textureSide$y = 256.0F;
```
Unless you plan on adding more stuff to that texture, shrink it (I think 128x64 should still fit everything)

Bernasss12 replied:
I think I'll keep it as is just because I don't know how much space I will need for the next step I'm thinking of taking 👍

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,316 @@
+		textY = topY + 24;
+		maxLines = (acrossY + borderSize) / 10;
+		/**Getting information from item NBT*/
+		if (hand == EnumHand.MAIN_HAND)
```
Use `player.getHeldItem(hand)` instead of the whole if block

dblsaiko commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -0,0 +1,76 @@
+	public void processPacket(EntityPlayer player) {
+		ItemStack book;
+		if (hand == EnumHand.MAIN_HAND) {
+			if (!(player.getHeldItemMainhand().getItem().getClass() == ItemGuideBook.class)) return;
```
Use `player.getHeldItem(hand)` instead of the whole if block

bziemons commented on common/logisticspipes/LPItems.java:
```diff
@@ -113,6 +114,9 @@
 	public static Item pipeFluidExtractor;
 
+	//working on it
+	@ObjectHolder("logisticspipes:guide_book")
```
Looking at the file, I think this would be better in the misc items group than between pipes. Also please remove the comment :)

Bernasss12 replied:
Yes that would indeed make the most sense, changed and comment removed :)

bziemons commented on common/logisticspipes/LogisticsPipes.java:
```diff
@@ -428,6 +429,9 @@ public void initItems(RegistryEvent.Register<Item> event) {
 		registry.register(setName(new ItemLogisticsChips(ItemLogisticsChips.ITEM_CHIP_FPGA), "chip_fpga"));
 		registry.register(setName(new ItemLogisticsChips(ItemLogisticsChips.ITEM_CHIP_FPGA_RAW), "chip_fpga_raw"));
 
+		//working on it
```
please remove :)

Bernasss12 replied:
Removed :)

bziemons commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,316 @@
+
+public class GuiGuideBook extends GuiScreen {
+
+	/**
```
double ** mark a JavaDoc comment, please use a single asterisk. The idea of this comment is good tho :+1:

Bernasss12 replied:
Surprisingly I didn't know the difference, but now I understand so I'll keep that in mind! Thanks!

Bernasss12 replied:
Hopefully I caught every instance I used it

bziemons commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,316 @@
+	private ItemStack book;
+	private EnumHand hand;
+
+	//Slider vars
```
Coding guidelines recommend a space between the start of a line comment and its content

Bernasss12 replied:
Corrected, I believe everywhere

bziemons commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,316 @@
+		this.hand = hand;
+	}
+
+	protected void splitLines() {
```
this does not just "split lines", but also saves them into the property `text`. Maybe instead return the list of split lines and pass `page` or better: the translated string, as an argument to it. Maybe the method can be a static function then, which does make more sense for a util function :)

Bernasss12 replied:
I did what I think you meant, please check in the next commit before resolving :)

bziemons replied:
sure thing :)

bziemons commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,316 @@
+		for (int i = splitSentences.size() - 1; i > 0; i--) {
+			splitSentences.add(i, "           ");
+		}
+		for (String sentence : splitSentences) {
```
Also if you like, have a look at the Java Streams API. It is quite straight forward once you get a hang of it

Bernasss12 replied:
What do you think would be the benefit of using that API? It would probably take me sometime to get the hang of a 'new' (for me) API

bziemons commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,316 @@
+
+	protected void drawForegroundLayer() {
+		GlStateManager.pushMatrix();
+		/** Drawing Foreground Border */
```
Again, JavaDoc comment which makes no sense here

bziemons replied:
wont mark further issues like this, please fix all of them :)

Bernasss12 replied:
FIxed all of them 👍

bziemons commented on resources/assets/logisticspipes/lang/en_us.lang:
```diff
@@ -514,3 +514,4 @@ book.quickstart.7=To flash a program, simply insert the Logistics Programmer int
 book.quickstart.8=You might be asking where the Mk2 and Mk3 pipes and modules are. They are gone, but you can recreate their effects by adding upgrades to your pipes and modules. Those upgrades are named 'Action Speed', 'Item Extraction' or 'ItemStack Extraction'.
 book.quickstart.9=To upgrade your modules, you need an upgrade for your chassis pipe called the 'Module Upgrade'. After applying it to your chassis pipe, new slots appear to the right of every inserted module. That's where the upgrades for the module go.
 book.quickstart.10=This 'quickstart' got longer than I anticipated, but I hope you got a good overview on how to start with this brand-new LP version. While you're at it you can also check out our new Highspeed Tubes. But most importantly: Have fun!
+book.quickstart.11=This 'quickstart' got longer than I anticipated, but I hope you got a good overview on how to start with this brand-new LP version. While you're at it you can also check out our new Highspeed Tubes. But most importantly: Have fun!This 'quickstart' got longer than I anticipated, but I hope you got a good overview on how to start with this brand-new LP version. While you're at it ...
```
delete this :)
I guess you could merge most pages together? Please test will all GUI sizes and different game window sizes though

Bernasss12 replied:
I did test that a lot and haven't been able to break it so far! :)
And also deleted!

bziemons reviewed (changes requested):
I see comments by @therealfarfetchd resolved although they aren't and it confuses me. If you have local changes that resolve those comments, please only mark them resolved here after pushing those changes.

Other than that I really like the looks of this and it will definitely be merged once it is done :)

Commit:
Formatting fixes

- Changed `splitLines()` method as requested :)
- Stopped using JavaDoc everywhere
- Formatted comments as they where supposed to be

Commit:
Manually Removed GuiGuideBook from GuiHandler

I thought I had reverted the class to the beginning of the branch, I guess it didn't work so there it goes :)

Bernasss12 commented on common/logisticspipes/network/packets/SetCurrentPagePacket.java:
```diff
@@ -0,0 +1,74 @@
+	public void processPacket(EntityPlayer player) {
+		ItemStack book;
+		book = player.getHeldItem(hand);
+		if (book.getItem() != LPItems.itemGuideBook) return;
```
@theZorro266 I actually had changed this as per @therealfarfetchd 's request but forgot to commit and push before going to bed, my bad :/

bziemons replied:
no problem

Bernasss12 commented:
Next thing on my list to add to this GuideBook would be a way to configure it without needing code, either via reading from a file, `.json` or otherwise, or by having some key strings in the `.lang` file, although this doesn't sound right to be honest...

To complement this addition I would like to implement a menu (also modular) so there's the possibility to add more than one line of thought, I.e.: "Quickstart Guide", "Basic Pipes" or "Basic Crafting System"

Another TODO I had in mind is to maybe have more than one page formats for example:
![image](https://user-images.githubusercontent.com/6233500/59342231-45f9e700-8d01-11e9-81fb-98e50222a853.png)
![image](https://user-images.githubusercontent.com/6233500/59342309-6fb30e00-8d01-11e9-9f69-dbdba7a83491.png)

dblsaiko commented on common/logisticspipes/network/GuiIDs.java:
```diff
@@ -27,6 +27,9 @@
 	 */
 
+	//Guide Book
+	public static final int GUI_Guide_Book_ID = 30;
```
unneeded after reverting GuiHandler

Bernasss12 replied:
Very true I'll remove that

dblsaiko commented on common/logisticspipes/proxy/side/ServerProxy.java:
```diff
@@ -249,6 +251,9 @@ public boolean checkSinglePlayerOwner(String commandSenderName) {
 	public void openFluidSelectGui(int slotId) {}
 
+	@Override
+	public void openGuiFromItem(int guiId, EnumHand hand) {}
```
Call this something more descriptive. `openGuideBookGui`?

Bernasss12 replied:
Very well it ended up not being id based so it doesn't make sense to have a general name

dblsaiko commented:
The simplest way to move the page contents out of the lang file (which is definitely a good idea, imo) is to create a seperate data structure in the assets directory, here's an example:
![guidebook](https://user-images.githubusercontent.com/3987560/59351267-48b70500-8d1e-11e9-8ae2-624819816be6.png)
This way you can have proper newlines and everything, and each page in a seperate file. Actually, the `page_count` in the book.json isn't really needed, since you can just try loading the next file and if that fails, the previous is your last page. But that also will prevent resource pack makers from making the book have fewer pages, so you should probably keep it if you go with this.

Bernasss12 commented:
Before any more additions this is in a good state to work with the language file, maybe after that we could think of another system to organise the book, imo

But if everyone agrees I like that file structure

Bernasss12 commented:
**So now I've added a texture, I hope it is to everyone's liking, if not change it however you would like :)**
![image](https://user-images.githubusercontent.com/6233500/59358930-5fad2580-8d25-11e9-8b60-adeae2a323e2.png)

#1242 
I also replaced the EventListener written book with this book. 
And added a recipe: 
![image](https://user-images.githubusercontent.com/6233500/59362278-195ac500-8d2b-11e9-807b-ed1892fcc1ea.png)

Commit:
Ready for Review

- Added Texture
- Added Recipe
- Added to EventListener
- Added localized name
- Removed from GuiIDs
- Renamed Proxy method
- Fixed minor bug

dblsaiko commented:
Texture looks good, looking through code later

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,331 @@
+				prevPageBtn.enabled = true;
+			}
+			if (page == pageMax) nextPageBtn.enabled = false;
+			else nextPageBtn.enabled = true;
```
`nextPageBtn.enabled = page != pageMax;`

Bernasss12 replied:
I did it this your way in `initGui()`, I don't even know why I made them like this here...

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,331 @@
+				nextPageBtn.enabled = true;
+			}
+			if (page == 1) prevPageBtn.enabled = false;
+			else prevPageBtn.enabled = true;
```
`prevPageBtn.enabled = page != 1;`

Bernasss12 replied:
Same as above :/

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,331 @@
+				sliderProgress = (slider.y - sliderTopY) / (float) (sliderBotY - sliderTopY);
+			} else {
+				slider.y = mouseY - slider.height / 2;
+				sliderProgress = (slider.y - sliderTopY) / (float) (sliderBotY - sliderTopY);
```
This line (`sliderProgress = …`) can be extracted out of all the if blocks since it is the same in all of them

Bernasss12 replied:
I agree, but this was replaced with the change below :)

dblsaiko commented on common/logisticspipes/gui/GuiGuideBook.java:
```diff
@@ -0,0 +1,331 @@
+	 */
+	protected void updateSlider() {
+		if (dragging && draggable) {
+			if (mouseY < sliderTopY + slider.height / 2) {
```
`mouseY < sliderTopY + slider.height / 2` = `mouseY - slider.height / 2 < sliderTopY`
`mouseY > sliderBotY + slider.height / 2` = `mouseY - slider.height / 2 > sliderBotY`
=> entire if/elseif chain can be replaced with `slider.y = MathHelper.clamp(mouseY - slider.height / 2, sliderTopY, sliderBotY);` after applying suggestion below (please also test that I got the params right)

Bernasss12 replied:
Works perfectly, very nice function that one!

dblsaiko commented on common/logisticspipes/items/ItemGuideBook.java:
```diff
@@ -0,0 +1,35 @@
+	@Override
+	public ActionResult<ItemStack> onItemRightClick(World world, EntityPlayer player, EnumHand hand) {
+		ItemStack itemstack = player.getHeldItem(hand);
+		if (world.isRemote) MainProxy.proxy.openGuideBookGui(hand);
```
With the proxy you technically don't need the isRemote check since it won't do anything on the server anyways, but I guess you can keep it

Bernasss12 replied:
I'll let it be then

Commit:
Applied Requested Changes, First Iteration

- Changed some parts of the code for some simpler equivalents.

bziemons commented:
whoever merges: please squash
